mayastor/src/subsys/nvmf/subsystem.rs

```diff
@@ -0,0 +1,561 @@
+                bdev.as_ptr(),
+                &opts as *const _,
+                size_of::<spdk_bdev_nvme_opts>() as u64,
+                ptr::null_mut(),
+            )
+        };
+
+        // the first name space should be 1 and we do not (currently) use
+        // more then one namespace
+
+        if ns_id < 1 {
+            Err(Error::Namespace {
+                bdev: bdev.name(),
+                msg: "failed to add namespace ID".to_string(),
+            })
+        } else {
+            info!("added NS ID {}", ns_id);
+            Ok(())
+        }
+    }
+
+    /// destroy the subsystem
+    pub fn destroy(&self) {
+        unsafe { spdk_nvmf_subsystem_destroy(self.0.as_ptr()) }
+    }
+
+    /// Get NVMe subsystem's NQN
+    pub fn get_nqn(&self) -> String {
+        unsafe {
+            spdk_nvmf_subsystem_get_nqn(self.0.as_ptr())
+                .as_str()
+                .to_string()
+        }
+    }
+    /// allow any host to connect to the subsystem
+    pub fn allow_any(&self, enable: bool) {
+        unsafe {
+            spdk_nvmf_subsystem_set_allow_any_host(self.0.as_ptr(), enable)
+        };
+    }
+
+    // we currently allow all listeners to the subsystem
+    async fn add_listener(&self) -> Result<(), Error> {
+        extern "C" fn listen_cb(arg: *mut c_void, status: i32) {
+            let s = unsafe { Box::from_raw(arg as *mut oneshot::Sender<i32>) };
+            s.send(status).unwrap();
+        }
+
+        let cfg = Config::get();
+
+        // dont yet enable both ports, IOW just add one transportID now
+
+        let trid_replica = TransportID::new(cfg.nexus_opts.nvmf_replica_port);
+
+        let (s, r) = oneshot::channel::<i32>();
+        unsafe {
+            spdk_nvmf_subsystem_add_listener(
+                self.0.as_ptr(),
+                trid_replica.as_ptr(),
+                Some(listen_cb),
+                cb_arg(s),
+            );
+        }
+
+        r.await.expect("listen a callback gone").to_result(|e| {
+            Error::Transport {
+                source: Errno::from_i32(e),
+                msg: "Failed to add listener".to_string(),
+            }
+        })
+    }
+
+    /// start the subsystem previously created -- note that we destroy it on
+    /// failure to ensure the state is not in limbo and to avoid leaking
+    /// resources
+    pub async fn start(self) -> Result<(), Error> {
+        extern "C" fn start_cb(
+            ss: *mut spdk_nvmf_subsystem,
+            arg: *mut c_void,
+            status: i32,
+        ) {
+            let s = unsafe { Box::from_raw(arg as *mut oneshot::Sender<i32>) };
+            let ss = NvmfSubsystem::from(ss);
+            if status != 0 {
+                error!(
+                    "Failed start subsystem state {} -- destroying it",
+                    ss.get_nqn()
+                );
+                ss.destroy();
+            }
+
+            s.send(status).unwrap();
+        }
+
+        self.add_listener().await?;
```

Comment:
Maybe, good question. Although it should not be able to fail. When we constructed the target -- we already registered the transport so we are roughly saying: that TCP transport you have? this subsystem should also be available over that, there are only a few reasons why this can fail but it can be retried again without doing any harm.

IOW, the subsystem will stay in whatever state it was prior to calling start if the add listener fails.